<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>38 系統樹 | 疫学のための R ハンドブック</title>
<meta name="author" content="the handbook team">
<meta name="description" content="38.1 概要 系統樹とは、生物の遺伝暗号の配列に基づいて、その近縁性や進化を可視化・記述するためのものです。...">
<meta name="generator" content="bookdown 0.29 with bs4_book()">
<meta property="og:title" content="38 系統樹 | 疫学のための R ハンドブック">
<meta property="og:type" content="book">
<meta property="og:description" content="38.1 概要 系統樹とは、生物の遺伝暗号の配列に基づいて、その近縁性や進化を可視化・記述するためのものです。...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="38 系統樹 | 疫学のための R ハンドブック">
<meta name="twitter:description" content="38.1 概要 系統樹とは、生物の遺伝暗号の配列に基づいて、その近縁性や進化を可視化・記述するためのものです。...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.0/transition.js"></script><script src="libs/bs3compat-0.4.0/tabs.js"></script><script src="libs/bs3compat-0.4.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.25/datatables.js"></script><link href="libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="libs/tabwid-1.1.0/tabwid.css" rel="stylesheet">
<link href="libs/tabwid-1.1.0/scrool.css" rel="stylesheet">
<link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.1.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.1.1/leaflet-providers-plugin.js"></script><script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.9/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.10.0/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script><link href="libs/vis-9.1.0/vis-network.min.css" rel="stylesheet">
<script src="libs/vis-9.1.0/vis-network.min.js"></script><script src="libs/visNetwork-binding-2.1.0/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-2.5.1/plotly-basic-2.5.1.min.js"></script><script src="libs/plotly-cartesian-2.5.1/plotly-cartesian-2.5.1.min.js"></script><script src="libs/proj4js-2.3.15/proj4.js"></script><link href="libs/highcharts-9.3.1/css/motion.css" rel="stylesheet">
<script src="libs/highcharts-9.3.1/highcharts.js"></script><script src="libs/highcharts-9.3.1/highcharts-3d.js"></script><script src="libs/highcharts-9.3.1/highcharts-more.js"></script><script src="libs/highcharts-9.3.1/modules/stock.js"></script><script src="libs/highcharts-9.3.1/modules/map.js"></script><script src="libs/highcharts-9.3.1/modules/data.js"></script><script src="libs/highcharts-9.3.1/modules/exporting.js"></script><script src="libs/highcharts-9.3.1/modules/offline-exporting.js"></script><script src="libs/highcharts-9.3.1/modules/drilldown.js"></script><script src="libs/highcharts-9.3.1/modules/item-series.js"></script><script src="libs/highcharts-9.3.1/modules/overlapping-datalabels.js"></script><script src="libs/highcharts-9.3.1/modules/annotations.js"></script><script src="libs/highcharts-9.3.1/modules/export-data.js"></script><script src="libs/highcharts-9.3.1/modules/funnel.js"></script><script src="libs/highcharts-9.3.1/modules/heatmap.js"></script><script src="libs/highcharts-9.3.1/modules/treemap.js"></script><script src="libs/highcharts-9.3.1/modules/sankey.js"></script><script src="libs/highcharts-9.3.1/modules/dependency-wheel.js"></script><script src="libs/highcharts-9.3.1/modules/organization.js"></script><script src="libs/highcharts-9.3.1/modules/solid-gauge.js"></script><script src="libs/highcharts-9.3.1/modules/streamgraph.js"></script><script src="libs/highcharts-9.3.1/modules/sunburst.js"></script><script src="libs/highcharts-9.3.1/modules/vector.js"></script><script src="libs/highcharts-9.3.1/modules/wordcloud.js"></script><script src="libs/highcharts-9.3.1/modules/xrange.js"></script><script src="libs/highcharts-9.3.1/modules/tilemap.js"></script><script src="libs/highcharts-9.3.1/modules/venn.js"></script><script src="libs/highcharts-9.3.1/modules/gantt.js"></script><script src="libs/highcharts-9.3.1/modules/timeline.js"></script><script src="libs/highcharts-9.3.1/modules/parallel-coordinates.js"></script><script src="libs/highcharts-9.3.1/modules/bullet.js"></script><script src="libs/highcharts-9.3.1/modules/coloraxis.js"></script><script src="libs/highcharts-9.3.1/modules/dumbbell.js"></script><script src="libs/highcharts-9.3.1/modules/lollipop.js"></script><script src="libs/highcharts-9.3.1/modules/series-label.js"></script><script src="libs/highcharts-9.3.1/plugins/motion.js"></script><script src="libs/highcharts-9.3.1/custom/reset.js"></script><script src="libs/highcharts-9.3.1/modules/boost.js"></script><script src="libs/highchart-binding-0.9.4/highchart.js"></script><script type="text/javascript">
        // window.addEventListener("load", function() {
        //     document.getElementById('toc').firstChild.innerHTML = "章目次"; 
        // });
        document.addEventListener("DOMContentLoaded", function() {
            document.getElementsByClassName("d-flex align-items-start justify-content-between")[0].children[0].children[0].innerHTML = "疫学のための R</br>ハンドブック";
            document.getElementById('toc').firstChild.innerHTML = "章目次"; 
            document.getElementById('main-nav').children[1].firstChild.innerHTML = "総目次";
            document.getElementById('search').setAttribute("placeholder", "検索");
        });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">疫学のための R ハンドブック</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"></a></li>
<li class="book-part">本書について</li>
<li><a class="" href="editorial-style.html"><span class="header-section-number">1</span> 編集前記・技術注記</a></li>
<li><a class="" href="data-used.html"><span class="header-section-number">2</span> ハンドブックとデータのダウンロード</a></li>
<li class="book-part">基本編</li>
<li><a class="" href="basics.html"><span class="header-section-number">3</span> R の基礎</a></li>
<li><a class="" href="transition-to-R.html"><span class="header-section-number">4</span> Excel・Stata・SASとの比較</a></li>
<li><a class="" href="packages-suggested.html"><span class="header-section-number">5</span> 推奨パッケージ</a></li>
<li><a class="" href="r-projects.html"><span class="header-section-number">6</span> R プロジェクト</a></li>
<li><a class="" href="importing.html"><span class="header-section-number">7</span> データのインポート・エクスポート</a></li>
<li class="book-part">データマネジメント</li>
<li><a class="" href="cleaning.html"><span class="header-section-number">8</span> データクリーニングと主要関数</a></li>
<li><a class="" href="dates.html"><span class="header-section-number">9</span> 日付型データ</a></li>
<li><a class="" href="characters-strings.html"><span class="header-section-number">10</span> 文字型・文字列型データ</a></li>
<li><a class="" href="factors.html"><span class="header-section-number">11</span> 因子（ファクタ）型データ</a></li>
<li><a class="" href="pivoting.html"><span class="header-section-number">12</span> データの縦横変換</a></li>
<li><a class="" href="grouping.html"><span class="header-section-number">13</span> データのグループ化</a></li>
<li><a class="" href="joining-matching.html"><span class="header-section-number">14</span> データの結合</a></li>
<li><a class="" href="deduplication.html"><span class="header-section-number">15</span> 重複データの排除</a></li>
<li><a class="" href="iteration.html"><span class="header-section-number">16</span> ループと反復処理・リストの操作</a></li>
<li class="book-part">データ分析</li>
<li><a class="" href="tables-descriptive.html"><span class="header-section-number">17</span> 記述統計表の作り方</a></li>
<li><a class="" href="stat-tests.html"><span class="header-section-number">18</span> 基本的な統計的検定</a></li>
<li><a class="" href="regression.html"><span class="header-section-number">19</span> 単変量と多変量回帰</a></li>
<li><a class="" href="missing-data.html"><span class="header-section-number">20</span> データの欠損</a></li>
<li><a class="" href="standardization.html"><span class="header-section-number">21</span> 標準化率</a></li>
<li><a class="" href="moving-average.html"><span class="header-section-number">22</span> 移動平均</a></li>
<li><a class="" href="time-series.html"><span class="header-section-number">23</span> 時系列分析とアウトブレイクの検出</a></li>
<li><a class="" href="epidemic-models.html"><span class="header-section-number">24</span> エピデミックモデリング</a></li>
<li><a class="" href="contact-tracing.html"><span class="header-section-number">25</span> 接触者の追跡</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">26</span> 標本調査データ分析</a></li>
<li><a class="" href="survival-analysis.html"><span class="header-section-number">27</span> 生存時間解析</a></li>
<li><a class="" href="gis.html"><span class="header-section-number">28</span> GIS の基本</a></li>
<li class="book-part">データの可視化</li>
<li><a class="" href="tables-presentation.html"><span class="header-section-number">29</span> 見やすい表の作り方</a></li>
<li><a class="" href="ggplot-basics.html"><span class="header-section-number">30</span> ggplot の基本</a></li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">31</span> ggplotのヒント</a></li>
<li><a class="" href="epicurves.html"><span class="header-section-number">32</span> 流行曲線（エピカーブ）</a></li>
<li><a class="" href="age-pyramid.html"><span class="header-section-number">33</span> 人口ピラミッドとリッカート尺度</a></li>
<li><a class="" href="heatmaps.html"><span class="header-section-number">34</span> ヒートマップ</a></li>
<li><a class="" href="diagrams.html"><span class="header-section-number">35</span> フローチャート・サンキー図・タイムライン</a></li>
<li><a class="" href="combination-analysis.html"><span class="header-section-number">36</span> 複数回答データの分析</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">37</span> 感染連鎖</a></li>
<li><a class="active" href="phylogenetic-trees.html"><span class="header-section-number">38</span> 系統樹</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">39</span> 動的な図の作成</a></li>
<li class="book-part">レポートとダッシュボード</li>
<li><a class="" href="rmarkdown.html"><span class="header-section-number">40</span> R Markdown で作るレポート</a></li>
<li><a class="" href="reportfactory.html"><span class="header-section-number">41</span> ルーチンで作成するレポートの整理</a></li>
<li><a class="" href="flexdashboard.html"><span class="header-section-number">42</span> R Markdownで作るダッシュボード</a></li>
<li><a class="" href="shiny-basics.html"><span class="header-section-number">43</span> Shiny で作るダッシュボード</a></li>
<li class="book-part">その他</li>
<li><a class="" href="writing-functions.html"><span class="header-section-number">44</span> 関数の作成</a></li>
<li><a class="" href="directories.html"><span class="header-section-number">45</span> ディレクトリの操作</a></li>
<li><a class="" href="collaboration.html"><span class="header-section-number">46</span> Git と Github を使ったバージョン管理と共同作業</a></li>
<li><a class="" href="errors.html"><span class="header-section-number">47</span> よくあるエラー</a></li>
<li><a class="" href="help.html"><span class="header-section-number">48</span> ヘルプ</a></li>
<li><a class="" href="network-drives.html"><span class="header-section-number">49</span> ネットワークドライブで R を使用する場合</a></li>
<li><a class="" href="data-table.html"><span class="header-section-number">50</span> data.table パッケージを使用したデータ処理</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenetic-trees" class="section level1" number="38">
<h1>
<span class="header-section-number">38</span> 系統樹<a class="anchor" aria-label="anchor" href="#phylogenetic-trees"><i class="fas fa-link"></i></a>
</h1>
<!-- ======================================================= -->
<div id="概要-6" class="section level2" number="38.1">
<h2>
<span class="header-section-number">38.1</span> 概要<a class="anchor" aria-label="anchor" href="#%E6%A6%82%E8%A6%81-6"><i class="fas fa-link"></i></a>
</h2>
<p><strong>系統樹</strong>とは、生物の遺伝暗号の配列に基づいて、その近縁性や進化を可視化・記述するためのものです。</p>
<p>系統樹は、遺伝子配列から、距離に基づく方法（近傍結合法など）や、特徴に基づく方法（最尤法やベイズ・マルコフ連鎖モンテカルロ法など）を用いて構築することができます。次世代シーケンシング（NGS）は、価格が手頃になり、感染症の原因となる病原体を記述するために公衆衛生の分野で広く使われるようになってきました。携帯型のシーケンサーを使用することで、解析時間を短縮し、リアルタイムでアウトブレイク調査に必要なデータを得ることができます。NGS データは、発生した菌株の起源や発生源、その繁殖を特定したり、抗菌剤耐性遺伝子の有無を判定したりするのに利用できます。また、サンプル間の遺伝的関連性を可視化するために、系統樹を構築します。</p>
<p>このページでは、系統樹とデータフレーム形式の追加サンプルデータを組み合わせて可視化することができる <strong>ggtree</strong> パッケージの使い方を学びます。これにより、パターンを観察し、アウトブレイクのダイナミックな理解を深めることができます。</p>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_overview_graph-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- ======================================================= -->
</div>
<div id="準備-29" class="section level2" number="38.2">
<h2>
<span class="header-section-number">38.2</span> 準備<a class="anchor" aria-label="anchor" href="#%E6%BA%96%E5%82%99-29"><i class="fas fa-link"></i></a>
</h2>
<div id="パッケージの読み込み-19" class="section level3 unnumbered">
<h3>パッケージの読み込み<a class="anchor" aria-label="anchor" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%AE%E8%AA%AD%E3%81%BF%E8%BE%BC%E3%81%BF-19"><i class="fas fa-link"></i></a>
</h3>
<p>このコードチャンクは、分析に必要なパッケージの読み込みを示しています。このハンドブックでは <strong>pacman</strong> の <code><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load()</a></code> を重視しています。<code><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load()</a></code> は必要に応じてパッケージをインストールし、使用するためにパッケージを読み込みます。インストールされたパッケージは R の <strong>base</strong> パッケージの <code><a href="https://rdrr.io/r/base/library.html">library()</a></code> でも読み込みできます。R のパッケージに関する詳細は <a href="#basic">R の基礎</a> の章をご覧ください。</p>
<div class="sourceCode" id="cb1727"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span></span>
<span>  <span class="va">rio</span>,             <span class="co"># インポート／エクスポート</span></span>
<span>  <span class="va">here</span>,            <span class="co"># 相対的なファイルパス</span></span>
<span>  <span class="va">tidyverse</span>,       <span class="co"># 一般的なデータマネジメントと可視化</span></span>
<span>  <span class="va">ape</span>,             <span class="co"># 系統樹ファイルのインポートとエクスポート</span></span>
<span>  <span class="va">ggtree</span>,          <span class="co"># 系統樹の可視化</span></span>
<span>  <span class="va">treeio</span>,          <span class="co"># 系統樹の可視化</span></span>
<span>  <span class="va">ggnewscale</span><span class="op">)</span>      <span class="co"># 色に関するレイヤーの追加</span></span></code></pre></div>
</div>
<div id="データのインポート-15" class="section level3 unnumbered">
<h3>データのインポート<a class="anchor" aria-label="anchor" href="#%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E3%82%A4%E3%83%B3%E3%83%9D%E3%83%BC%E3%83%88-15"><i class="fas fa-link"></i></a>
</h3>
<p>本章で扱われるデータは、<a href="data-used.html#data-used">ハンドブックとデータのダウンロード</a> の章の説明に従ってダウンロードできます。</p>
<p>系統樹の保存形式にはいくつかの種類があります（ Newick、NEXUS、Phylip など）。一般的なのはNewickファイル形式（.nwk）で、これは木をコンピュータで読める形で表現するための標準的なものです。つまり、木全体を “((t2:0.04,t1:0.34):0.89,(t5:0.37,(t4:0.03,t3:0.67):0.9):0.59);” のような文字列形式で表現し、すべてのノードと先端、そしてそれらの関係（枝の長さ）を列挙することができます。</p>
<p>注： 系統樹ファイル自体にはシーケンスデータは含まれておらず、単に配列間の遺伝的距離の結果であることを理解しておくことが重要です。したがって、ツリーファイルからシーケンスデータを抽出することはできません。</p>
<p>まず、<strong>ape</strong> パッケージの <code>read.tree()</code> 関数を使って、Newick の系統樹ファイルを .txt 形式で読み込み、“phylo” 型のリストオブジェクトに格納しておきます。必要に応じて、<strong>here</strong> パッケージの <code><a href="https://here.r-lib.org//reference/here.html">here()</a></code> 関数を使用して、相対ファイルパスを指定します。</p>
<p>注：このケースでは、Github からのダウンロードや取り扱いを容易にするために、newick ツリーを .txt ファイルとして保存しています。</p>
<div class="sourceCode" id="cb1728"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="st">"Shigella_tree.txt"</span><span class="op">)</span></span></code></pre></div>
<p>tree オブジェクトを調べてみると、299 個の先端（またはサンプル）と 236 個のノードが含まれています。</p>
<div class="sourceCode" id="cb1729"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span></span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 299 tips and 236 internal nodes.
## 
## Tip labels:
##   SRR5006072, SRR4192106, S18BD07865, S18BD00489, S17BD08906, S17BD05939, ...
## Node labels:
##   17, 29, 100, 67, 100, 100, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>次に、<strong>rio</strong> パッケージの <code><a href="https://rdrr.io/pkg/rio/man/import.html">import()</a></code> 関数を用いて、性別、原産国、抗菌薬への薬剤耐性など、シーケンスされた各サンプルの追加情報を含む .csv ファイルとして保存された表をインポートします。</p>
<div class="sourceCode" id="cb1731"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rio/man/import.html">import</a></span><span class="op">(</span><span class="st">"sample_data_Shigella_tree.csv"</span><span class="op">)</span></span></code></pre></div>
<p>以下は、データの最初の 50 行です。</p>
<div id="htmlwidget-2855021fad10ea78aefe" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2855021fad10ea78aefe">{"x":{"filter":"none","vertical":false,"data":[["S17BD05944","S15BD07413","S18BD07247","S19BD07384","S18BD07338","S18BD02657","S19BD06884","S16BD02574","S16BD04069","S17BD08170","S15BD07250","S17BD07935","S17BD00384","S19BD06034","S17BD08948","S17BD04542","S17BD07399","S17BD08233","S19BD07381","S17BD07961","S18BD09176","S19BD07393","S18BD02580","S17BD02526","S17BD06833","S18BD01024","S18BD03411","S18BD09515","S17BD07332","S18BD06598","S17BD02164","S19BD06386","S19BD06283","S19BD06285","S19BD05221","S18BD06749","S17BD02088","S17BD07692","S14BD03573","S17BD01135","S19BD06689","S15BD06969","S18BD01106","S17BD00525","S19BD07036","S18BD06401","S18BD06486","S17BD06426","S17BD05916","S18BD07708"],["Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei"],["Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium"],["Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe"],[null,"Georgia",null,"Morocco",null,null,null,"Thailand",null,null,null,null,null,null,null,"Iran",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,"Morocco","France","Turkey","Egypt","Tunisia","Congo",null,null,null,null,null,null,null,null,null,null,"Egypt",null,null],["2017","2015","2018","2019","2018","2018","2019","2016","2016","2017","2015","2017","2017","2019","2017","2017","2017","2017","2019","2017","2018","2019","2018","2017","2017","2018","2018","2018","2017","2018","2017","2019","2019","2019","2019","2018","2017","2017","2014","2017","2019","2015","2018","2017","2019","2018","2018","2017","2017","2018"],["Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes"],["NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL"],[null,"Male","Female","Female","Male","Male","Male","Male","Male","Male","Female",null,"Male","Male","Female","Male","Male","Female","Female","Male","Male","Male","Male","Female","Male","Female","Female","Male","Female","Male","Female","Female","Female","Female","Female","Female","Male","Male",null,"female","Male","Female","Male","Male","Male","Male","Male","Male","Male","Female"],[null,null,null,null,null,"gyrA D87Y","gyrA D87Y","gyrA D87Y","gyrA D87Y",null,null,"gyrA D87Y",null,null,"gyrA D87Y",null,null,null,null,null,null,null,null,null,null,"gyrA D87Y","gyrA S83L",null,null,null,null,null,null,null,null,null,null,"gyrA D87Y","gyrA D87Y",null,null,null,"gyrA D87Y",null,null,null,null,null,null,null],[null,null,null,null,null,"mphA","mphA",null,null,null,null,"mphA",null,null,"mphA",null,null,null,null,null,null,null,null,null,null,"mphA","mphA",null,null,null,null,null,null,null,"mphA/ermB",null,null,"mphA",null,null,null,null,"mphA",null,null,null,null,null,null,null],[8,0,16,8,8,64,64,0,0,64,0,64,8,8,64,8,8,8,8,8,8,16,8,8,8,64,64,8,8,8,8,8,8,2,64,8,8,64,0,8,8,0,64,8,4,8,8,8,8,8],[0.25,0,0.03,0.02,0.02,0.5,0.5,0,0,8,0,0.5,0.02,0.02,0.5,0.25,0.02,0.02,0.02,0.02,0.02,0.12,0.02,0.02,0.02,0.5,0.25,0.03,0.02,0.02,0.02,0.02,0.02,0.02,0.02,0.02,0.02,1,0,0.02,0.02,0,0.5,0.02,0.02,4,0.02,0.02,0.25,0.02]],"container":"<table class=\"white-space: nowrap\">\n  <thead>\n    <tr>\n      <th>Sample_ID<\/th>\n      <th>serotype<\/th>\n      <th>Country<\/th>\n      <th>Continent<\/th>\n      <th>Travel_history<\/th>\n      <th>Year<\/th>\n      <th>Belgium<\/th>\n      <th>Source<\/th>\n      <th>Gender<\/th>\n      <th>gyrA_mutations<\/th>\n      <th>macrolide_resistance_genes<\/th>\n      <th>MIC_AZM<\/th>\n      <th>MIC_CIP<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[11,12]}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="クリーニングと点検" class="section level3 unnumbered">
<h3>クリーニングと点検<a class="anchor" aria-label="anchor" href="#%E3%82%AF%E3%83%AA%E3%83%BC%E3%83%8B%E3%83%B3%E3%82%B0%E3%81%A8%E7%82%B9%E6%A4%9C"><i class="fas fa-link"></i></a>
</h3>
<p>データのクリーニングと点検を行います。正しいサンプルデータを系統樹に割り当てるためには、<code>sample_data</code> データフレームの <code>Sample_ID</code> カラムの値が、<code>tree</code> ファイルの <code>tip.labels</code> の値と一致する必要があります。</p>
<p>R の <strong>base</strong> パッケージの <code><a href="https://rdrr.io/r/utils/head.html">head()</a></code> を使って最初の 6 つのエントリーを見ることで、 <code>tree</code> ファイルの <code>tip.labels</code> のフォーマットをチェックします。</p>
<div class="sourceCode" id="cb1732"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span> </span></code></pre></div>
<pre><code>## [1] "SRR5006072" "SRR4192106" "S18BD07865" "S18BD00489" "S17BD08906"
## [6] "S17BD05939"</code></pre>
<p>また、<code>sample_data</code> データフレームの最初の列が <code>Sample_ID</code> であることを確認します。R の <strong>base</strong> パッケージの <code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> を使って、データフレームのカラム名を調べます。</p>
<div class="sourceCode" id="cb1734"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">)</span>   </span></code></pre></div>
<pre><code>##  [1] "Sample_ID"                  "serotype"                  
##  [3] "Country"                    "Continent"                 
##  [5] "Travel_history"             "Year"                      
##  [7] "Belgium"                    "Source"                    
##  [9] "Gender"                     "gyrA_mutations"            
## [11] "macrolide_resistance_genes" "MIC_AZM"                   
## [13] "MIC_CIP"</code></pre>
<p>データフレーム内の <code>Sample_IDs</code> を見て、フォーマットが <code>tip.label</code> と同じであることを確認します（例：文字はすべて大文字、文字と数字の間に余分なアンダースコア <code>_</code> がない、など）。</p>
<div class="sourceCode" id="cb1736"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">)</span> <span class="co"># もう一度、head() を使用して、初めの 6 つのデータだけ確認する</span></span></code></pre></div>
<pre><code>## [1] "S17BD05944" "S15BD07413" "S18BD07247" "S19BD07384" "S18BD07338"
## [6] "S18BD02657"</code></pre>
<p>また、すべてのサンプルが <code>tree</code> ファイルに存在するかどうか、あるいはその逆かどうかを、一致する部分と一致しない部分に TRUE または FALSE のロジカルベクトルを生成することで比較することができます。これらは、簡単にするために、ここでは表示しません。</p>
<div class="sourceCode" id="cb1738"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span></span>
<span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></span></code></pre></div>
<p>これらのベクトルを使って、 tree 上にないサンプル ID を表示することができます（結果として、そのような例はありません）。</p>
<div class="sourceCode" id="cb1739"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">[</span><span class="op">!</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">]</span></span></code></pre></div>
<pre><code>## character(0)</code></pre>
<p>調べてみると、データフレーム内の <code>Sample_ID</code> のフォーマットが、<code>tip.labels</code> のサンプル名のフォーマットに対応していることがわかります。これらは、同じ順序でソートされていなくてもマッチします。</p>
<p>これで準備完了です。</p>
<!-- ======================================================= -->
</div>
</div>
<div id="単純な系統樹の可視化" class="section level2" number="38.3">
<h2>
<span class="header-section-number">38.3</span> 単純な系統樹の可視化<a class="anchor" aria-label="anchor" href="#%E5%8D%98%E7%B4%94%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E5%8F%AF%E8%A6%96%E5%8C%96"><i class="fas fa-link"></i></a>
</h2>
<div id="様々な系統樹のレイアウト" class="section level3 unnumbered">
<h3>様々な系統樹のレイアウト<a class="anchor" aria-label="anchor" href="#%E6%A7%98%E3%80%85%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E3%83%AC%E3%82%A4%E3%82%A2%E3%82%A6%E3%83%88"><i class="fas fa-link"></i></a>
</h3>
<p><strong>ggtree</strong> には様々なレイアウトが用意されており、目的に応じて最適なものを選ぶことができます。以下にいくつかのデモンストレーションを紹介します。他のオプションについては、この<a href="http://yulab-smu.top/treedata-book/chapter4.html">オンラインブック</a>を参照してください。</p>
<p>以下は、ツリーレイアウトの例です。</p>
<div class="sourceCode" id="cb1741"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>                                            <span class="co"># 単純な線形の系統樹</span></span>
<span><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>,  branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>                   <span class="co"># 全ての先端が揃えられた単純な線形の系統樹</span></span>
<span><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span>                         <span class="co"># 単純な円形の系統樹</span></span>
<span><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span>, branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="co"># 全ての先端が揃えられた単純な円形の系統樹</span></span></code></pre></div>
<p><img src="_main_files/figure-html/phylogenetic_trees_example_formats-1.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-2.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-3.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-4.png" width="50%"></p>
</div>
<div id="単純な系統樹とサンプルデータ" class="section level3 unnumbered">
<h3>単純な系統樹とサンプルデータ<a class="anchor" aria-label="anchor" href="#%E5%8D%98%E7%B4%94%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%A8%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB%E3%83%87%E3%83%BC%E3%82%BF"><i class="fas fa-link"></i></a>
</h3>
<p><code>sample_data</code> データフレームを <code>tree</code> ファイルに接続するには、<strong>%&lt;+%</strong> 演算子を使用します。
系統樹の最も簡単な注釈は、先端にサンプル名を追加することと、先端のポイントや必要に応じて枝に色を付けることです。</p>
<p>以下は、円形の系統樹の例です。</p>
<div class="sourceCode" id="cb1742"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout <span class="op">=</span> <span class="st">"circular"</span>, branch.length <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span> <span class="co"># %&lt;+% で、tree にサンプルデータを追加する</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">Belgium</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>                       <span class="co"># データフレームの変数に応じて分枝に色付ける</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Sample Origin"</span>,                      <span class="co"># カラースキームの名前</span></span>
<span>    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span><span class="op">)</span>,                     <span class="co"># 変数の異なるオプション</span></span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NRCSS Belgium"</span>, <span class="st">"Other"</span><span class="op">)</span>,        <span class="co"># 凡例の中で異なるオプションをどのように命名するか</span></span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"black"</span><span class="op">)</span>,                  <span class="co"># 変数に割り当てたい色</span></span>
<span>    na.value <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>                        <span class="co">#  NA 値を黒に</span></span>
<span>  <span class="fu">new_scale_color</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>                             <span class="co"># 別の変数の配色を追加する</span></span>
<span>    <span class="fu">geom_tippoint</span><span class="op">(</span></span>
<span>      mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Continent</span><span class="op">)</span>,          <span class="co"># 大陸別の先端の色。"shape = " を加えて形状を変更する</span></span>
<span>      size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span>                               <span class="co"># 先端のポイントのサイズを定義する</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Continent"</span>,                    <span class="co"># カラースキームの名前（凡例ではこのように表示される</span></span>
<span>    palette <span class="op">=</span> <span class="st">"Set1"</span>,                      <span class="co"># brewer パッケージに付属しているカラーセットを選ぶ</span></span>
<span>    na.value <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span>                    <span class="co"># NA 値はグレーを選択</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>                             <span class="co"># 枝の先端にサンプルの名前を追加 </span></span>
<span>    color <span class="op">=</span> <span class="st">'black'</span>,                       <span class="co"># (テキストラインは + で好きなだけ追加できるが、隣り合うように配置するにはオフセット値を調整する必要がある)</span></span>
<span>    offset <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    geom <span class="op">=</span> <span class="st">"text"</span>,</span>
<span>    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span>    </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Phylogenetic tree of Shigella sonnei"</span><span class="op">)</span><span class="op">+</span>       <span class="co"># グラフのタイトル</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, <span class="co"># x 軸のタイトルを削除</span></span>
<span>    axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, <span class="co"># y 軸のタイトルを削除</span></span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>    <span class="co"># 凡例のタイトルのフォントサイズとフォーマットを定義する</span></span>
<span>      face <span class="op">=</span> <span class="st">"bold"</span>,</span>
<span>      size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,   </span>
<span>    legend.text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>       <span class="co"># 凡例テキストのフォントサイズとフォーマットを定義する</span></span>
<span>      face <span class="op">=</span> <span class="st">"bold"</span>,</span>
<span>      size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,  </span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>      <span class="co"># プロットタイトルのフォントサイズとフォーマットを定義する</span></span>
<span>      size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>      face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,  </span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>,     <span class="co"># 凡例の配置を決める</span></span>
<span>    legend.box <span class="op">=</span> <span class="st">"vertical"</span>,        <span class="co"># 凡例の配置を決める</span></span>
<span>    legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>   </span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_adding_sampledata-1.png" width="672" style="display: block; margin: auto;"></div>
<p>他の ggplot オブジェクトと同じように、<code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave()</a></code> を使ってツリープロットをエクスポートすることができます。このように書くと、<code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave()</a></code> は最後に生成された画像を指定されたファイルパスに保存します。<code><a href="https://here.r-lib.org//reference/here.html">here()</a></code> や相対ファイルパスを使えば、サブフォルダなどにも簡単に保存できることを覚えておいてください。</p>
<div class="sourceCode" id="cb1743"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="st">"example_tree_circular_1.png"</span>, width <span class="op">=</span> <span class="fl">12</span>, height <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></span></code></pre></div>
<!-- ======================================================= -->
</div>
</div>
<div id="系統樹の操作処理加工" class="section level2" number="38.4">
<h2>
<span class="header-section-number">38.4</span> 系統樹の操作・処理・加工<a class="anchor" aria-label="anchor" href="#%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E6%93%8D%E4%BD%9C%E5%87%A6%E7%90%86%E5%8A%A0%E5%B7%A5"><i class="fas fa-link"></i></a>
</h2>
<p>非常に大きな系統樹であっても、その中の一部分にしか興味がない場合があります。例えば、歴史的または国際的なサンプルを含む系統樹を作成して、データセットが全体像の中でどこに当てはまるかを大まかに把握する場合です。しかし、データをより詳しく見るためには、大きな系統樹の一部だけを確認したいとします。</p>
<p>系統樹ファイルはシーケンシングデータの解析結果に過ぎないので、ファイル内のノードや枝の順番を操作することはできません。これらは、生の NGS データから以前の解析ですでに決定されています。しかし、部分的に拡大したり、部分的に非表示にしたり、系統樹の一部をサブセットにしたりすることは可能です。</p>
<div id="拡大" class="section level3 unnumbered">
<h3>拡大<a class="anchor" aria-label="anchor" href="#%E6%8B%A1%E5%A4%A7"><i class="fas fa-link"></i></a>
</h3>
<p>系統樹を “切断” するのではなく、一部だけをより詳しく調べたい場合は、ズームインして特定の部分を表示することができます。</p>
<p>まず、系統樹全体を線形のフォーマットでプロットし、系統樹の各ノードに数字のラベルを追加します。</p>
<div class="sourceCode" id="cb1744"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>,<span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span>                <span class="co"># すべての枝の先端に、tree ファイルのサンプル名をラベル付ける</span></span>
<span>  <span class="fu">geom_text2</span><span class="op">(</span></span>
<span>    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">!</span><span class="va">isTip</span>,</span>
<span>                  label <span class="op">=</span> <span class="va">node</span><span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"darkred"</span>,</span>
<span>    hjust <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>                            <span class="co"># 系統樹内のすべてのノードにラベルを付ける</span></span>
<span></span>
<span><span class="va">p</span>  <span class="co"># 表示する</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_zoom_in-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>ある特定の枝（右に突き出ている）にズームインするには、ggtree オブジェクト <code>p</code> で <code>viewClade()</code> を使用し、ノード番号を指定すると、より詳しく見ることができます。</p>
<div class="sourceCode" id="cb1745"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">viewClade</span><span class="op">(</span><span class="va">p</span>, node <span class="op">=</span> <span class="fl">452</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_zoom_in_452-1.png" width="50%" style="display: block; margin: auto;"></div>
</div>
<div id="枝の折り畳み" class="section level3 unnumbered">
<h3>枝の折り畳み<a class="anchor" aria-label="anchor" href="#%E6%9E%9D%E3%81%AE%E6%8A%98%E3%82%8A%E7%95%B3%E3%81%BF"><i class="fas fa-link"></i></a>
</h3>
<p>しかし、この枝を無視したい場合もあるので、<code><a href="https://dplyr.tidyverse.org/reference/compute.html">collapse()</a></code> を使って同じノード（ノード番号 452）で枝を折り畳むことができます。このツリーは <code>p_collapsed</code> として定義されます。</p>
<div class="sourceCode" id="cb1746"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_collapsed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html">collapse</a></span><span class="op">(</span><span class="va">p</span>, node <span class="op">=</span> <span class="fl">452</span><span class="op">)</span></span>
<span><span class="va">p_collapsed</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_collapse_452-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>分かりやすくするために、<code>p_collapsed</code> をプリントする際に、折り畳まれた枝のノードに <code>geom_point2()</code>（青い菱形）を追加しています。</p>
<div class="sourceCode" id="cb1747"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_collapsed</span> <span class="op">+</span> </span>
<span><span class="fu">geom_point2</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">(</span><span class="va">node</span> <span class="op">==</span> <span class="fl">452</span><span class="op">)</span><span class="op">)</span>,  <span class="co"># 折り畳んだノードに記号を割り当てる</span></span>
<span>            size <span class="op">=</span> <span class="fl">5</span>,                     <span class="co"># 記号のサイズを定義する</span></span>
<span>            shape <span class="op">=</span> <span class="fl">23</span>,                   <span class="co"># 記号の形を定義する</span></span>
<span>            fill <span class="op">=</span> <span class="st">"steelblue"</span><span class="op">)</span>           <span class="co"># 記号の塗りつぶしを定義する</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-1319-1.png" width="672"></div>
</div>
<div id="系統樹の分割サブセット" class="section level3 unnumbered">
<h3>系統樹の分割（サブセット）<a class="anchor" aria-label="anchor" href="#%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E5%88%86%E5%89%B2%E3%82%B5%E3%83%96%E3%82%BB%E3%83%83%E3%83%88"><i class="fas fa-link"></i></a>
</h3>
<p>より永続的な変更を行い、作業用の縮小された新しい系統樹を作成したい場合は、 <code>tree_subset()</code> でツリーの一部をサブセットします。そして、それを新しい newick tree ファイルか .txt ファイルとして保存します。</p>
<div class="sourceCode" id="cb1748"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggtree</span><span class="op">(</span></span>
<span>  <span class="va">tree</span>,</span>
<span>  branch.length <span class="op">=</span> <span class="st">'none'</span>,</span>
<span>  layout <span class="op">=</span> <span class="st">'circular'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>               <span class="co"># %&lt;+% 演算子を使って、サンプルデータを追加する</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">+</span>                                <span class="co"># すべての枝の先端にサンプル名をラベル付けしてツリーファイルに保存</span></span>
<span>  <span class="fu">geom_text2</span><span class="op">(</span></span>
<span>    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">!</span><span class="va">isTip</span>, label <span class="op">=</span> <span class="va">node</span><span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span> <span class="op">+</span>                                <span class="co"># 系統樹内のすべてのノードにラベルを付ける</span></span>
<span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>   legend.position <span class="op">=</span> <span class="st">"none"</span>,                            <span class="co"># 凡例も全て削除</span></span>
<span>   axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>   axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>   plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_subsetting-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>ここで、ノード 528 で系統樹をサブセットする（ノード 528 以降の枝の中の先端だけを残す）ことにして、それを新しい <code>sub_tree1</code> オブジェクトとして保存したとします。</p>
<div class="sourceCode" id="cb1749"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sub_tree1</span> <span class="op">&lt;-</span> <span class="fu">tree_subset</span><span class="op">(</span></span>
<span>  <span class="va">tree</span>,</span>
<span>  node <span class="op">=</span> <span class="fl">528</span><span class="op">)</span>                                            <span class="co"># ノード 528 で系統樹をサブセットする</span></span></code></pre></div>
<p>subset tree 1 を見てみましょう。</p>
<div class="sourceCode" id="cb1750"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Subset tree 1"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-1321-1.png" width="672"></div>
<p>また、1 つの特定のサンプルに基づいてサブセットすることもできます。その際には、含める「後ろ」のノード数を指定します。系統樹の同じ部分を、サンプル（ここでは S17BD07692）に基づいて 9 ノードさかのぼってサブセットし、新しい <code>sub_tree2</code> オブジェクトとして保存します。</p>
<div class="sourceCode" id="cb1751"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sub_tree2</span> <span class="op">&lt;-</span> <span class="fu">tree_subset</span><span class="op">(</span></span>
<span>  <span class="va">tree</span>,</span>
<span>  <span class="st">"S17BD07692"</span>,</span>
<span>  levels_back <span class="op">=</span> <span class="fl">9</span><span class="op">)</span> <span class="co"># levels back は、先端（各サンプル）から何ノード後方に移動するかを定義する</span></span></code></pre></div>
<p>subset tree 2 を見てみましょう。</p>
<div class="sourceCode" id="cb1752"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">3</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Subset tree 2"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-1323-1.png" width="672"></div>
<p>また、<strong>ape</strong> パッケージの <code>write.tree()</code> 関数を使って、新しい系統樹を Newick 型やテキストファイルとして保存することもできます。</p>
<div class="sourceCode" id="cb1753"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># .nwk フォーマットで保存</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html">write.tree</a></span><span class="op">(</span><span class="va">sub_tree2</span>, file<span class="op">=</span><span class="st">'data/phylo/Shigella_subtree_2.nwk'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># .txt フォーマットで保存</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html">write.tree</a></span><span class="op">(</span><span class="va">sub_tree2</span>, file<span class="op">=</span><span class="st">'data/phylo/Shigella_subtree_2.txt'</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="系統樹内でのノードの回転" class="section level3 unnumbered">
<h3>系統樹内でのノードの回転<a class="anchor" aria-label="anchor" href="#%E7%B3%BB%E7%B5%B1%E6%A8%B9%E5%86%85%E3%81%A7%E3%81%AE%E3%83%8E%E3%83%BC%E3%83%89%E3%81%AE%E5%9B%9E%E8%BB%A2"><i class="fas fa-link"></i></a>
</h3>
<p>前述したように、ツリー内の先端やノードの順序を変更することはできません。これは、遺伝子の関連性に基づいており、視覚的な操作の対象にはならないからです。しかし、視覚的にわかりやすくするために、ノードの周りに枝を回転させることはできます。</p>
<p>まず、ノードのラベルとともに新しい subset tree 2 をプロットして、操作したいノードを選び、それを ggtree plot オブジェクト <code>p</code> に格納します。</p>
<div class="sourceCode" id="cb1754"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">+</span>  </span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_text2</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span>, label<span class="op">=</span><span class="va">node</span><span class="op">)</span>, <span class="co"># 系統樹内のすべてのノードにラベルを付ける</span></span>
<span>             size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>             color <span class="op">=</span> <span class="st">"darkred"</span>, </span>
<span>             hjust <span class="op">=</span> <span class="fl">1</span>, </span>
<span>             vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span>
<span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_rotating_1-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>その後、 <strong>ggtree::rotate()</strong> や <strong>ggtree::flip()</strong> を使ってノードを操作することができます。
注：どのノードを操作しているかを示すために、まず、<strong>ggtree</strong> の <strong>geom_hilight()</strong> 関数を適用して、興味のあるノードのサンプルをハイライトし、その ggtree プロットオブジェクトを新しいオブジェクト <code>p1</code> に格納します。</p>
<div class="sourceCode" id="cb1755"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">geom_hilight</span><span class="op">(</span>  <span class="co"># ノード 39 を青くハイライトし、"extend =" でカラーブロックの長さを定義する</span></span>
<span>  node <span class="op">=</span> <span class="fl">39</span>,</span>
<span>  fill <span class="op">=</span> <span class="st">"steelblue"</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0.0017</span><span class="op">)</span> <span class="op">+</span>  </span>
<span><span class="fu">geom_hilight</span><span class="op">(</span>            <span class="co"># ノード 37 を黄色くハイライトする</span></span>
<span>  node <span class="op">=</span> <span class="fl">37</span>,</span>
<span>  fill <span class="op">=</span> <span class="st">"yellow"</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0.0017</span><span class="op">)</span> <span class="op">+</span>               </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Original tree"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p1</span> <span class="co"># 表示する</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_rotating_2-1.png" width="50%" style="display: block; margin: auto;"></div>
<p>ここで、オブジェクト <code>p1</code> のノード 37 を回転させ、ノード 38 上のサンプルが一番上に移動するようにします。回転したツリーを新しいオブジェクト <code>p2</code> に格納します。</p>
<div class="sourceCode" id="cb1756"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">rotate</span><span class="op">(</span><span class="va">p1</span>, <span class="fl">37</span><span class="op">)</span> <span class="op">+</span> </span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Rotated Node 37"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p2</span>   <span class="co"># 表示する</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-1324-1.png" width="672"></div>
<p>また、flip コマンドを使って、オブジェクト <code>p1</code> のノード 36 を回転させ、ノード 37 を上に、ノード 39 を下に切り替えることもできます。反転したツリーを新しいオブジェクト <code>p3</code> に格納します。</p>
<div class="sourceCode" id="cb1757"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p3</span> <span class="op">&lt;-</span>  <span class="fu">flip</span><span class="op">(</span><span class="va">p1</span>, <span class="fl">39</span>, <span class="fl">37</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Rotated Node 36"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p3</span>   <span class="co"># 表示する</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/unnamed-chunk-1325-1.png" width="672"></div>
</div>
<div id="サンプルデータを付加したサブツリーの例" class="section level3 unnumbered">
<h3>サンプルデータを付加したサブツリーの例<a class="anchor" aria-label="anchor" href="#%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB%E3%83%87%E3%83%BC%E3%82%BF%E3%82%92%E4%BB%98%E5%8A%A0%E3%81%97%E3%81%9F%E3%82%B5%E3%83%96%E3%83%84%E3%83%AA%E3%83%BC%E3%81%AE%E4%BE%8B"><i class="fas fa-link"></i></a>
</h3>
<p>サブツリーのノード 39 で 2017 年と 2018 年に発生したクローン拡大を伴う症例のクラスタを調査しているとします。他の近縁種の株の起源を見るために、株の分離年に加えて、渡航歴と国別の色を追加します。</p>
<div class="sourceCode" id="cb1758"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>     <span class="co"># %&lt;+% 演算子を使って sample_data にリンクしている</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>                          <span class="co"># すべての枝の先端に、ツリーファイルのサンプル名をラベル付けする</span></span>
<span>    size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>    offset <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_tree2</span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.015</span><span class="op">)</span><span class="op">+</span>                       <span class="co"># 系統樹の x 軸の 下限、上限を設定</span></span>
<span>  <span class="fu">geom_tippoint</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">Country</span><span class="op">)</span>,     <span class="co"># 大陸別に先端を色付ける</span></span>
<span>                size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Country"</span>, </span>
<span>    palette <span class="op">=</span> <span class="st">"Set1"</span>, </span>
<span>    na.value <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>                          <span class="co"># 隔離された年をテキストラベルとして先端に追加</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Year</span><span class="op">)</span>,</span>
<span>    color <span class="op">=</span> <span class="st">'blue'</span>,</span>
<span>    offset <span class="op">=</span> <span class="fl">0.0045</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    linetype <span class="op">=</span> <span class="st">"blank"</span> ,</span>
<span>    geom <span class="op">=</span> <span class="st">"text"</span>,</span>
<span>    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span> </span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>                          <span class="co"># 先端のテキストラベルに渡航歴を赤で表示</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Travel_history</span><span class="op">)</span>,</span>
<span>    color <span class="op">=</span> <span class="st">'red'</span>,</span>
<span>    offset <span class="op">=</span> <span class="fl">0.006</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    linetype <span class="op">=</span> <span class="st">"blank"</span>,</span>
<span>    geom <span class="op">=</span> <span class="st">"text"</span>,</span>
<span>    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Phylogenetic tree of Belgian S. sonnei strains with travel history"</span><span class="op">)</span><span class="op">+</span>  <span class="co"># プロットタイトルの追加</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"genetic distance (0.001 = 4 nucleotides difference)"</span><span class="op">)</span><span class="op">+</span>                    <span class="co"># x 軸にラベルを付ける </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>    axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_inspect_subset_example-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>私たちの観察によると、アジアからの菌株の輸入があり、それが何年もかけてベルギーで伝播し、今回の流行を引き起こしたと考えられます。</p>
<!-- ======================================================= -->
</div>
</div>
<div id="より複雑な系統樹サンプルデータのヒートマップの追加" class="section level2 unnumbered">
<h2>より複雑な系統樹：サンプルデータのヒートマップの追加<a class="anchor" aria-label="anchor" href="#%E3%82%88%E3%82%8A%E8%A4%87%E9%9B%91%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E3%83%92%E3%83%BC%E3%83%88%E3%83%9E%E3%83%83%E3%83%97%E3%81%AE%E8%BF%BD%E5%8A%A0"><i class="fas fa-link"></i></a>
</h2>
<p><strong>ggtree::gheatmap()</strong> 関数を使用して、抗菌薬耐性遺伝子のカテゴリー的な存在や、実際に測定された抗菌薬耐性の数値など、より複雑な情報をヒートマップの形で追加することができます。</p>
<p>まず、系統樹をプロットし（これは線形でも円形でも構いません）、それを新しい ggtree plot のオブジェクト <code>p</code> に保存する必要があります。ここでは、パート 3 の sub_tree を使用します)。</p>
<div class="sourceCode" id="cb1759"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span>, branch.length<span class="op">=</span><span class="st">'none'</span>, layout<span class="op">=</span><span class="st">'circular'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>   legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span></span>
<span>      size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>      face <span class="op">=</span> <span class="st">"bold"</span>,</span>
<span>      hjust <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>      vjust <span class="op">=</span> <span class="op">-</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap-1.png" width="60%" style="display: block; margin: auto;"></div>
<p>次に、データの準備です。異なる変数を新しい配色で可視化するために、データフレームを目的の変数にサブセットします。行の名前として <code>Sample_ID</code> を追加することが重要で、そうしないとデータを <code>tip.labels</code> の系統樹にマッチさせることができません。</p>
<p>この例では、性別と、赤痢菌感染症の治療に使用される重要な第一選択抗菌薬であるシプロフロキサシンへの耐性を与える可能性のある変異を調べたいと思います。</p>
<p>性別のデータフレームを作成します。</p>
<div class="sourceCode" id="cb1760"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gender</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"gender"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gender"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gender</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></span></code></pre></div>
<p>シプロフロキサシン耐性をもたらす gyrA 遺伝子の変異について、データフレームを作成します。</p>
<div class="sourceCode" id="cb1761"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cipR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"cipR"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gyrA_mutations"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cipR</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></span></code></pre></div>
<p>実験室で測定されたシプロフロキサシンの最小発育阻止濃度（MIC）のデータフレームを作成します。</p>
<div class="sourceCode" id="cb1762"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MIC_Cip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"mic_cip"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MIC_CIP"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">MIC_Cip</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></span></code></pre></div>
<p>系統樹に性別の二値のヒートマップを追加した最初のプロットを作成し、それを新しい ggtree plot のオブジェクト <code>h1</code> に格納します。</p>
<div class="sourceCode" id="cb1763"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h1</span> <span class="op">&lt;-</span>  <span class="fu">gheatmap</span><span class="op">(</span><span class="va">p</span>, <span class="va">gender</span>,                                 <span class="co"># 性別データフレームのヒートマップレイヤーをツリープロットに追加する</span></span>
<span>                offset <span class="op">=</span> <span class="fl">10</span>,                               <span class="co"># offset は、ヒートマップを右方向に移動する</span></span>
<span>                width <span class="op">=</span> <span class="fl">0.10</span>,                              <span class="co"># width は、ヒートマップのカラムの幅を定義する</span></span>
<span>                color <span class="op">=</span> <span class="cn">NULL</span>,                              <span class="co"># color は、ヒートマップの列の境界線を定義する</span></span>
<span>         colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>                               <span class="co"># ヒートマップのカラム名を隠す</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>,                       <span class="co"># 性別の配色と凡例の定義</span></span>
<span>                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#00d1b1"</span>, <span class="st">"purple"</span><span class="op">)</span>,</span>
<span>                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>,</span>
<span>                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will
## replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb1766"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h1</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_gender-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>その後、シプロフロキサシンに対する耐性を付与する gyrA 遺伝子の変異の情報を追加しています。</p>
<p>注：WGS データにおける染色体の点変異の有無は、Zankari らが開発した PointFinder ツールを用いて事前に判断しています（参考文献の項を参照）。</p>
<p>まず、既存のプロットオブジェクト <code>h1</code> に新しい配色を割り当てて、今のオブジェクト <code>h2</code> に格納します。これにより、ヒートマップの 2 つ目の変数の色を定義・変更することができます。</p>
<div class="sourceCode" id="cb1767"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p>次に、2 つ目のヒートマップレイヤーを <code>h2</code> に追加し、合成したプロットを新しいオブジェクト <code>h3</code> に格納します。</p>
<div class="sourceCode" id="cb1768"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h3</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h2</span>, <span class="va">cipR</span>,         <span class="co"># ヒートマップの 2 行目に Ciprofloxacin の耐性変異を追加</span></span>
<span>               offset <span class="op">=</span> <span class="fl">12</span>, </span>
<span>               width <span class="op">=</span> <span class="fl">0.10</span>, </span>
<span>               colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Ciprofloxacin resistance \n conferring mutation"</span>,</span>
<span>                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#fe9698"</span>,<span class="st">"#ea0c92"</span><span class="op">)</span>,</span>
<span>                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA D87Y"</span>, <span class="st">"gyrA S83L"</span><span class="op">)</span>,</span>
<span>                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA d87y"</span>, <span class="st">"gyrA s83l"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">2</span>,byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will
## replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb1771"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h3</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_cip_genes-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>以上のプロセスを繰り返し、まず既存のオブジェクト <code>h3</code> に新しいカラースケール層を追加し、次に得られたオブジェクト <code>h4</code> に各菌株に対するシプロフロキサシンの最小発育阻止濃度（MIC）の連続データを追加して、最終的なオブジェクト <code>h5</code> を作成します。</p>
<div class="sourceCode" id="cb1772"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># まず、新しいカラーリングを追加する</span></span>
<span><span class="va">h4</span> <span class="op">&lt;-</span> <span class="va">h3</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># そして、その 2 つを組み合わせて新しいプロットを作る</span></span>
<span><span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h4</span>, <span class="va">MIC_Cip</span>,  </span>
<span>               offset <span class="op">=</span> <span class="fl">14</span>, </span>
<span>               width <span class="op">=</span> <span class="fl">0.10</span>,</span>
<span>                colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html">scale_fill_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"MIC for Ciprofloxacin"</span>,  <span class="co"># ここでは、MIC の連続変数のグラデーションカラースキームを定義する</span></span>
<span>                      low <span class="op">=</span> <span class="st">"yellow"</span>, high <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                      breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.50</span>, <span class="fl">1.00</span><span class="op">)</span>,</span>
<span>                      na.value <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_colourbar.html">guide_colourbar</a></span><span class="op">(</span>barwidth <span class="op">=</span> <span class="fl">5</span>, barheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will
## replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb1775"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h5</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_cip_MIC-1.png" width="90%" style="display: block; margin: auto;"></div>
<p>これと同じことを、線形の系統樹についても行うことができます。</p>
<div class="sourceCode" id="cb1776"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> <span class="co"># 先端にラベル付け</span></span>
<span>  <span class="fu">theme_tree2</span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"genetic distance (0.001 = 4 nucleotides difference)"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.015</span><span class="op">)</span><span class="op">+</span></span>
<span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>      axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, </span>
<span>                                face <span class="op">=</span> <span class="st">"bold"</span>,</span>
<span>                                hjust <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                vjust <span class="op">=</span> <span class="op">-</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_1-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>まず、性別を追加します。</p>
<div class="sourceCode" id="cb1777"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h1</span> <span class="op">&lt;-</span>  <span class="fu">gheatmap</span><span class="op">(</span><span class="va">p</span>, <span class="va">gender</span>, </span>
<span>                offset <span class="op">=</span> <span class="fl">0.003</span>,</span>
<span>                width <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>                color<span class="op">=</span><span class="st">"black"</span>, </span>
<span>         colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Gender"</span>,</span>
<span>                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#00d1b1"</span>, <span class="st">"purple"</span><span class="op">)</span>,</span>
<span>                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>,</span>
<span>                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will
## replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb1780"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h1</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_2-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>そして、もう一つの配色レイヤーを追加した後に、シプロフロキサシン耐性変異を追加します。</p>
<div class="sourceCode" id="cb1781"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">h3</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h2</span>, <span class="va">cipR</span>,   </span>
<span>               offset <span class="op">=</span> <span class="fl">0.004</span>, </span>
<span>               width <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>               color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Ciprofloxacin resistance \n conferring mutation"</span>,</span>
<span>                    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#fe9698"</span>,<span class="st">"#ea0c92"</span><span class="op">)</span>,</span>
<span>                    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA D87Y"</span>, <span class="st">"gyrA S83L"</span><span class="op">)</span>,</span>
<span>                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA d87y"</span>, <span class="st">"gyrA s83l"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        legend.box <span class="op">=</span> <span class="st">"vertical"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">2</span>,byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will
## replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb1784"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">h3</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_3-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>そして、実験室で決められた最小発育阻止濃度（MIC）を加えます。</p>
<div class="sourceCode" id="cb1785"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h4</span> <span class="op">&lt;-</span> <span class="va">h3</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span><span class="va">h4</span>, <span class="va">MIC_Cip</span>, </span>
<span>               offset <span class="op">=</span> <span class="fl">0.005</span>,  </span>
<span>               width <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>               color <span class="op">=</span> <span class="st">"black"</span>, </span>
<span>                colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html">scale_fill_continuous</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"MIC for Ciprofloxacin"</span>,</span>
<span>                      low <span class="op">=</span> <span class="st">"yellow"</span>, high <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                      breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.50</span>,<span class="fl">1.00</span><span class="op">)</span>,</span>
<span>                      na.value <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span><span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_colourbar.html">guide_colourbar</a></span><span class="op">(</span>barwidth <span class="op">=</span> <span class="fl">5</span>, barheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>        legend.box <span class="op">=</span> <span class="st">"horizontal"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Scale for 'y' is already present. Adding another scale for 'y', which will
## replace the existing scale.</code></pre>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb1788"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h5</span></span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear_4-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- ======================================================= -->
</div>
<div id="参考資料-29" class="section level2" number="38.5">
<h2>
<span class="header-section-number">38.5</span> 参考資料<a class="anchor" aria-label="anchor" href="#%E5%8F%82%E8%80%83%E8%B3%87%E6%96%99-29"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>
<a href="http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#" class="uri">http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#</a> Clade_Colors</li>
<li><a href="https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a></li>
<li><a href="https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html" class="uri">https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html</a></li>
<li><a href="https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a></li>
</ul>
<p>Ea Zankari, Rosa Allesøe, Katrine G Joensen, Lina M Cavaco, Ole Lund, Frank M Aarestrup, PointFinder: a novel web tool for WGS-based detection of antimicrobial resistance associated with chromosomal point mutations in bacterial pathogens, Journal of Antimicrobial Chemotherapy, Volume 72, Issue 10, October 2017, Pages 2764–2768, <a href="https://doi.org/10.1093/jac/dkx217" class="uri">https://doi.org/10.1093/jac/dkx217</a></p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="transmission-chains.html"><span class="header-section-number">37</span> 感染連鎖</a></div>
<div class="next"><a href="interactive-plots.html"><span class="header-section-number">39</span> 動的な図の作成</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenetic-trees"><span class="header-section-number">38</span> 系統樹</a></li>
<li><a class="nav-link" href="#%E6%A6%82%E8%A6%81-6"><span class="header-section-number">38.1</span> 概要</a></li>
<li>
<a class="nav-link" href="#%E6%BA%96%E5%82%99-29"><span class="header-section-number">38.2</span> 準備</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#%E3%83%91%E3%83%83%E3%82%B1%E3%83%BC%E3%82%B8%E3%81%AE%E8%AA%AD%E3%81%BF%E8%BE%BC%E3%81%BF-19">パッケージの読み込み</a></li>
<li><a class="nav-link" href="#%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E3%82%A4%E3%83%B3%E3%83%9D%E3%83%BC%E3%83%88-15">データのインポート</a></li>
<li><a class="nav-link" href="#%E3%82%AF%E3%83%AA%E3%83%BC%E3%83%8B%E3%83%B3%E3%82%B0%E3%81%A8%E7%82%B9%E6%A4%9C">クリーニングと点検</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#%E5%8D%98%E7%B4%94%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E5%8F%AF%E8%A6%96%E5%8C%96"><span class="header-section-number">38.3</span> 単純な系統樹の可視化</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#%E6%A7%98%E3%80%85%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E3%83%AC%E3%82%A4%E3%82%A2%E3%82%A6%E3%83%88">様々な系統樹のレイアウト</a></li>
<li><a class="nav-link" href="#%E5%8D%98%E7%B4%94%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%A8%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB%E3%83%87%E3%83%BC%E3%82%BF">単純な系統樹とサンプルデータ</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E6%93%8D%E4%BD%9C%E5%87%A6%E7%90%86%E5%8A%A0%E5%B7%A5"><span class="header-section-number">38.4</span> 系統樹の操作・処理・加工</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#%E6%8B%A1%E5%A4%A7">拡大</a></li>
<li><a class="nav-link" href="#%E6%9E%9D%E3%81%AE%E6%8A%98%E3%82%8A%E7%95%B3%E3%81%BF">枝の折り畳み</a></li>
<li><a class="nav-link" href="#%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%81%AE%E5%88%86%E5%89%B2%E3%82%B5%E3%83%96%E3%82%BB%E3%83%83%E3%83%88">系統樹の分割（サブセット）</a></li>
<li><a class="nav-link" href="#%E7%B3%BB%E7%B5%B1%E6%A8%B9%E5%86%85%E3%81%A7%E3%81%AE%E3%83%8E%E3%83%BC%E3%83%89%E3%81%AE%E5%9B%9E%E8%BB%A2">系統樹内でのノードの回転</a></li>
<li><a class="nav-link" href="#%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB%E3%83%87%E3%83%BC%E3%82%BF%E3%82%92%E4%BB%98%E5%8A%A0%E3%81%97%E3%81%9F%E3%82%B5%E3%83%96%E3%83%84%E3%83%AA%E3%83%BC%E3%81%AE%E4%BE%8B">サンプルデータを付加したサブツリーの例</a></li>
</ul>
</li>
<li><a class="nav-link" href="#%E3%82%88%E3%82%8A%E8%A4%87%E9%9B%91%E3%81%AA%E7%B3%BB%E7%B5%B1%E6%A8%B9%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E3%83%92%E3%83%BC%E3%83%88%E3%83%9E%E3%83%83%E3%83%97%E3%81%AE%E8%BF%BD%E5%8A%A0">より複雑な系統樹：サンプルデータのヒートマップの追加</a></li>
<li><a class="nav-link" href="#%E5%8F%82%E8%80%83%E8%B3%87%E6%96%99-29"><span class="header-section-number">38.5</span> 参考資料</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>疫学のための R ハンドブック</strong>" was written by the handbook team. It was last built on 2022-10-04.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
